core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -580,7 +580,18 @@ abstract class RDD[T: ClassTag](
    * elements (a, b) where a is in `this` and b is in `other`.
    */
   def cartesian[U: ClassTag](other: RDD[U]): RDD[(T, U)] = withScope {
-    new CartesianRDD(sc, this, other)
+    val numPartitions = {
+      val numExecutors = System.getProperty("spark.executor.instances")
+
+      if (numExecutors != null) {
+        numExecutors.toInt
+      } else {
+        sc.defaultMinPartitions
+      }
+    }
+    val coalesced = coalesce(numPartitions, shuffle = true)
+    val coalescedOther = other.coalesce(numPartitions, shuffle = true)
+    new CartesianRDD(sc, coalesced, coalescedOther)
```

Comment:
Sorry I think I was vague when I originally brought up `coalesce`.  I had intended for it to only be a suggestion to the user, included in the docs, as you note.  I think its really hard to guess how to do the coalescing.  In fact, I just realized there is a downside to coalescing that I hadn't thought of earlier -- the more you coalesce, then it becomes more likely that you *cannot* store all of `rdd2.iterator` in memory.